### ISSUE TITLE
Document how to run the tests

### ISSUE THREAD

**Issue by zaquestion** [zaquestion/lab#239]
Accept that the test setup has factually grown complicated and its completely non-obvious how to run the tests.

**Comment by claytonrcarter** [comment]
Yes please. 😄 I'm happy to help with this, though as I've said elsewhere, I'm new to go and go testing but I'm not new to writing READMEs.

**Comment by zaquestion** [comment]
I'll say up front, theres room for simplification and some obvious improvements. To start and in the context of this issue, I'll focus on describing the current requirements which should help seed documentation (and some ideas for improvement)

Firstly, its important to note that lab run integration tests in addition to unittests. These integration tests execute against gitlab.com. We are willing to trade some test "reliability" in exchange for 100% guarantees that functionality works (because its run against live gitlab). Integration tests are largely identified as tests which execute the `./lab_bin` binary.
Ref on how lab achieves this can be seen in the TestMain and Makefile
- https://github.com/zaquestion/lab/blob/master/cmd/root_test.go#L25-L83
- https://github.com/zaquestion/lab/blob/master/Makefile#L10-L19

This means that a valid SSH key is require to execute a significant number of tests. This key must be attach to **a GitLab account**. Practically speaking this means a developer on lab must have a gitlab account with their ssh key attached.
- May be fixable by embeding the ssh key used by travis in the test harness. This would reduce the dependency to using an ssh agent (so that we could `echo <key> |  ssh-add -`
- Alternatively spin up gitlab in docker and populate with test objects.

Nextly, tests must be run via `make test` to properly setup the `testdata/test.git`, the `-run` go flag is supported through `make test run=<>`
- Might be entirely fixable by using a git submodule. This was a feature I was scared of and very unfamiliar at the start of lab (so I copied hub's approach). I understand them now however and I believe the change would allow tests to be run with `go test` directly. This would also be useful because we could properly track gitlab.com/zaquestion/test instead of faking it.

Lastly some tests might still be broken (I think I've seen from coworkers), likely due to permissions issues I haven't surfaced. lab executes the tests with the lab-testing user which shouldn't have any permissions issues however. Documenting the above and addressing the outstanding warts would be a huge step forward. If there are lingering test failures, we'd at least be in a place to address them.

**Comment by claytonrcarter** [comment]
Sorry to need to so much hand holding on this. I'm blindly feeling my way through go and getting up and running. For the rank beginners, I just found this very helpful for getting up and running: https://golang.org/doc/code.html#Workspaces. I'm also finding that `GOPATH` and `GO111MODULES=on` environment vars need to be defined, is that correct? And `$GOPATH/bin` needs to be in your PATH.

After reading that, and setting those vars, I was able to run all of the tests successfully.
